Frameworks/Foundation/NSOperation.mm

```diff
@@ -92,14 +91,21 @@ - (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NS
             _finishCondition.notify_all();
             _finishLock.unlock();
         } else {
+            std::lock_guard<std::recursive_mutex> lock(_readyLock);
+            if ([object isFinished]) {
+                if ([_dependencies containsObject:object]) {
```

Comment:
This check is necessary to prevent a race condition between removeDependency and observeValueForKeyPath resulting from that dependency finishing. In removeDependency, we check the dependency's isFinished property and decrement _readyCount if that dependency is finished. If for some reason it is not finished when we check it, but it completes and calls observeValueForKeyPath before removeObserver is called, then observeValueForKeypath will block until _readyLock can be held, and at that point the dependency will no longer be contained in the _dependencies set, ensuring that _readyCount is not modified.

---
In reply to: [72368995](https://github.com/Microsoft/WinObjC/pull/691#discussion_r72368995) [](ancestors = 72368995)